src/routing/route.cc

```diff
@@ -190,31 +200,164 @@ LaneSRangeRelation Route::ComputeLaneSRangeRelation(const api::LaneSRange& lane_
+        break;
+
+      case LaneSRangeRelation::kRight:
+      case LaneSRangeRelation::kAdjacentRight:
+        current_lane_s_range_index.second += kTowardsLeft;
+        break;
+
+      case LaneSRangeRelation::kCoincident:
+        // Do nothing, leaving the loop now.
+        break;
+
+      default:
+        MALIPUT_THROW_MESSAGE("Code should not arrive here.");
+        break;
+    };
+  }
+  return api::LaneSRoute(lane_s_ranges);
+}
+
+std::optional<Route::LaneSRangeIndex> Route::FindLaneSRangeIndex(const api::LaneSRange& lane_s_range) const {
+  if (lane_id_to_lane_s_range_indices_.find(lane_s_range.lane_id()) == lane_id_to_lane_s_range_indices_.end()) {
+    return {};
+  }
+  const double tolerance = road_network_->road_geometry()->linear_tolerance();
+  const std::vector<LaneSRangeIndex> lane_s_range_indices = lane_id_to_lane_s_range_indices_.at(lane_s_range.lane_id());
+  for (const LaneSRangeIndex& idx : lane_s_range_indices) {
+    if (phases_[idx.first].lane_s_ranges()[idx.second].Contains(lane_s_range, tolerance)) {
+      return {idx};
+    }
   }
   return {};
 }
 
+std::optional<Route::LaneSRangeIndex> Route::FindStraightPredecessor(
+    const Route::LaneSRangeIndex& lane_s_range_index) const {
+  const size_t preceeding_phase_index = lane_s_range_index.first - 1u;
+  const Phase& preceeding_phase = phases_.at(preceeding_phase_index);
+  const api::LaneSRange& lane_s_range = GetLaneSRange(lane_s_range_index.first, lane_s_range_index.second);
+  const auto preceeding_lane_s_range_it =
+      std::find_if(preceeding_phase.lane_s_ranges().begin(), preceeding_phase.lane_s_ranges().end(),
+                   [this, lane_s_range](const api::LaneSRange& preceeding_lane_s_range) {
+                     return ComputeLaneSRangeRelation(lane_s_range, preceeding_lane_s_range) ==
+                            LaneSRangeRelation::kPreceedingStraight;
+                   });
+  if (preceeding_lane_s_range_it == preceeding_phase.lane_s_ranges().end()) {
+    return {};
+  }
+  const size_t preceeding_lane_s_range_index =
+      static_cast<size_t>(std::distance(preceeding_phase.lane_s_ranges().begin(), preceeding_lane_s_range_it));
+  return {LaneSRangeIndex{preceeding_phase_index, preceeding_lane_s_range_index}};
+}
+
+int Route::FindDirectionTowardsLaneSRangeWithStraightPredecessor(
+    const Route::LaneSRangeIndex& lane_s_range_index) const {
+  const Phase& current_phase = Get(lane_s_range_index.first);
+  const size_t min_current_lane_s_range_index = 0u;
+  const size_t max_current_lane_s_range_index = current_phase.lane_s_ranges().size() - 1u;
+
+  // Quick checks before looking for one side.
+  if (lane_s_range_index.second == min_current_lane_s_range_index) {  // This is the top right index, move to the left.
+    return kTowardsLeft;
+  } else if (lane_s_range_index.second ==
+             max_current_lane_s_range_index) {  // This is the top right index, move to the left.
```

Comment:
Is the comment correct? Should it be opposite?

```
// This is the top left index, move to the right.
```

I am not following what "top left/right index" means (see other comment regarding whether "left" and "right" are well defined). I simply noticed that this comment the same as the one on line 338, which seems strange.

Reply:
I think with the previous explanation this comment is satisfied, right?

Reply:
I'm still confused. This comment says "move to the left" yet the code below on line 349 returns `kTowardsRight`.